pkg/reconciler/workload/namespace/namespace_reconcile_scheduling.go

```diff
@@ -0,0 +1,243 @@
+			return reconcileStatusStop, ns, err
+		}
+
+		validPlacements = filterValidPlacements(ns, placements)
+	}
+
+	//1. pick all workload cluster in all bound placements
+	validClusters := []*workloadv1alpha1.WorkloadCluster{}
+	var errs []error
+	for _, placement := range validPlacements {
+		clusters, err := r.getAllValidWorkloadClustersForPlacement(clusterName, placement, ns)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if len(clusters) > 0 {
+			validClusters = append(validClusters, clusters...)
+		}
+	}
+
+	if len(errs) > 0 {
+		return reconcileStatusStop, ns, utilerrors.NewAggregate(errs)
+	}
+
+	// 2. find the scheduled workload cluster to the ns, including synced, removing
+	synced, removing := syncedRemovingCluster(ns)
+
+	// 3. if the synced cluster is in the valid clusters, stop scheduling
+	expectedAnnotations := map[string]interface{}{} // nil means to remove the key
+	expectedLabels := map[string]interface{}{}      // nil means to remove the key
+
+	validClusterSets := sets.NewString()
+	for _, wc := range validClusters {
+		validClusterSets.Insert(wc.Name)
+	}
+
+	syncedSets := sets.NewString(synced...)
+	for _, cluster := range synced {
+		if validClusterSets.Has(cluster) {
+			continue
+		}
+		// it is no longer a synced cluster, mark it as removing.
+		now := time.Now().UTC().Format(time.RFC3339)
+		expectedAnnotations[workloadv1alpha1.InternalClusterDeletionTimestampAnnotationPrefix+cluster] = now
+		syncedSets.Delete(cluster)
+		klog.V(4).Infof("set cluster %s removing for ns %s|%s since it is not a valid cluster anymore", cluster, clusterName, ns.Name)
+	}
+
+	// 4. if removing cluster is in the valid cluster, turns it into synced, otherwise, check if the removing cluster
+	// should be removed.
+	minEnqueueDuration := removingGracePeriod + 1
+	for cluster, removingTime := range removing {
+		if validClusterSets.Has(cluster) {
+			expectedLabels[workloadv1alpha1.InternalClusterResourceStateLabelPrefix+cluster] = string(workloadv1alpha1.ResourceStateSync)
+			expectedAnnotations[workloadv1alpha1.InternalClusterDeletionTimestampAnnotationPrefix+cluster] = nil
+			syncedSets.Insert(cluster)
+			klog.V(4).Infof("turn cluster %s from removing to sync for ns %s|%s since it is now a valid cluster", cluster, clusterName, ns.Name)
+			continue
+		}
+
+		if removingTime.Add(removingGracePeriod).Before(time.Now()) {
+			expectedLabels[workloadv1alpha1.InternalClusterResourceStateLabelPrefix+cluster] = nil
+			expectedAnnotations[workloadv1alpha1.InternalClusterDeletionTimestampAnnotationPrefix+cluster] = nil
+			klog.V(4).Infof("remove cluster %s for ns %s|%s", cluster, clusterName, ns.Name)
+		} else {
+			enqueuDuration := time.Until(removingTime.Add(removingGracePeriod))
+			if enqueuDuration < minEnqueueDuration {
+				minEnqueueDuration = enqueuDuration
+			}
+		}
+	}
+
+	// 5. randomly select a cluster if there is no cluster syncing currently.
+	if syncedSets.Len() == 0 && len(validClusters) > 0 {
```

Comment:
🔴  this surprises. We have to go location by location, and select one for each, no?